src/Symfony/Component/DependencyInjection/Secrets/JweEntry.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace Symfony\Component\DependencyInjection\Secrets;
+
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+final class JweEntry
+{
+    //TODO do we need to store header and auth tag separately?
+    private $header;
+    private $encryptedKey;
+    private $iv;
+    private $cipherText;
+    private $authTag;
+
+    public static function decrypt(string $compactedEntry, string &$keyPair)
+    {
+        $entry = new self();
+        $entry->hydrate($compactedEntry);
+        $plaintext = $entry->decryptedSecret($keyPair);
+        sodium_memzero($keyPair);
+
+        return $plaintext;
+    }
+
+    public static function encrypt(string $secret, string $pubKey)
+    {
+        $entry = new self();
+        $entry->setEncryptedValues($secret, $pubKey);
+
+        return $entry->compact();
+    }
+
+    private function base64url_encode($data)
+    {
+        return urlencode(base64_encode($data));
+    }
+
+    private function base64url_decode($data)
+    {
+        return base64_decode(urldecode($data));
+    }
+
+    // see the JWE compact serialization format for details https://tools.ietf.org/html/rfc7516#section-7.1
+    private function compact()
+    {
+        $valuesForCompaction = array(
+            $this->base64url_encode($this->header),
+            sodium_bin2base64($this->encryptedKey, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->iv, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->cipherText, SODIUM_BASE64_VARIANT_URLSAFE),
+            $this->base64url_encode($this->authTag),
+        );
+
+        return implode('.', $valuesForCompaction);
+    }
+
+    private function decryptedSecret(string &$keyPair)
+    {
+        $plaintextCek = sodium_crypto_box_seal_open(
+            $this->encryptedKey,
+            $keyPair
+        );
+
+        $decrypted = sodium_crypto_aead_aes256gcm_decrypt(
+            $this->cipherText,
+            $this->authTag,
+            $this->iv,
+            $plaintextCek
+        );
+        if (false === $decrypted) {
+            throw new RuntimeException(sprintf(
+                'Unable to decrypt secrets. Verify the configured key pair'
+            ));
+        }
+
+        return $decrypted;
+    }
+
+    private function generateHeader()
+    {
+        //TODO: can you force utf8 in string declaration in php?
+        //TODO: figure out correct name for alg
+        return '{"alg":"curve25519xsalsa20poly1305","enc":"A256GCM"}';
```

Comment:
These still look like important TODOs... and I don't understand what any of this means of where it came from - that should definitely get a comment.